presto-main/src/main/java/com/facebook/presto/operator/scalar/Centroid.java

```diff
@@ -27,130 +27,165 @@
+    private int count;
 
     // The ID is transient because it must be unique within a given JVM. A new
     // ID should be generated from uniqueCount when a Centroid is deserialized.
     private transient int id;
 
-    private List<Double> actualData = null;
+    private List<Double> actualData;
 
-    private Centroid(boolean record) {
+    private Centroid(boolean record)
+    {
         id = uniqueCount.getAndIncrement();
         if (record) {
             actualData = new ArrayList<>();
         }
     }
 
-    public Centroid(double x) {
+    public Centroid(double value)
+    {
         this(false);
-        start(x, 1, uniqueCount.getAndIncrement());
+        start(value, 1, uniqueCount.getAndIncrement());
     }
 
-    public Centroid(double x, int w) {
+    public Centroid(double value, int weight)
+    {
         this(false);
-        start(x, w, uniqueCount.getAndIncrement());
+        start(value, weight, uniqueCount.getAndIncrement());
     }
 
-    public Centroid(double x, int w, int id) {
+    public Centroid(double value, int weight, int id)
+    {
         this(false);
-        start(x, w, id);
+        start(value, weight, id);
     }
 
-    public Centroid(double x, int id, boolean record) {
+    public Centroid(double value, int id, boolean record)
+    {
         this(record);
-        start(x, 1, id);
+        start(value, 1, id);
     }
 
-    Centroid(double x, int w, List<Double> data) {
-        this(x, w);
+    Centroid(double value, int weight, List<Double> data)
+    {
+        this(value, weight);
         actualData = data;
     }
 
-    private void start(double x, int w, int id) {
+    private void start(double value, int weight, int id)
+    {
         this.id = id;
-        add(x, w);
+        add(value, weight);
     }
 
-    public void add(double x, int w) {
+    public void add(double value, int weight)
+    {
         if (actualData != null) {
-            actualData.add(x);
+            actualData.add(value);
         }
-        count += w;
-        centroid += w * (x - centroid) / count;
+        count += weight;
+        mean += weight * (value - mean) / count;
     }
 
-    public double mean() {
-        return centroid;
+    public double mean()
+    {
+        return mean;
     }
 
-    public int count() {
+    public int count()
+    {
         return count;
     }
 
-    public int id() {
+    public int id()
+    {
         return id;
     }
 
     @Override
-    public String toString() {
+    public String toString()
+    {
         return "Centroid{" +
-                "centroid=" + centroid +
+                "mean=" + mean +
                 ", count=" + count +
                 '}';
     }
 
     @Override
-    public int hashCode() {
+    public int hashCode()
+    {
         return id;
     }
 
     @Override
-    public int compareTo(@SuppressWarnings("NullableProblems") Centroid o) {
-        int r = Double.compare(centroid, o.centroid);
+    public boolean equals(Object o)
```

Comment:
It's a bit strange to see `hashCode` and `equals` not based on the same variables. What's the reason behind this?

Reply:
Each instance of a centroid has a unique id, even if we copy the centroid containing the same values. Therefore, if we compare by id, it would always return false.